internal/logger/logger.go

```diff
@@ -230,46 +232,12 @@ func selectComponentLevels(componentLevels map[Component]Level) map[Component]Le
 	return selected
 }
 
-// truncate will truncate a string to the given width, appending "..." to the
-// end of the string if it is truncated. This routine is safe for multi-byte
-// characters.
-func truncate(str string, width uint) string {
-	if width == 0 {
-		return ""
-	}
-
-	if len(str) <= int(width) {
-		return str
-	}
-
-	// Truncate the byte slice of the string to the given width.
-	newStr := str[:width]
-
-	// Check if the last byte is at the beginning of a multi-byte character.
-	// If it is, then remove the last byte.
-	if newStr[len(newStr)-1]&0xC0 == 0xC0 {
-		return newStr[:len(newStr)-1] + TruncationSuffix
-	}
-
-	// Check if the last byte is in the middle of a multi-byte character. If
-	// it is, then step back until we find the beginning of the character.
-	if newStr[len(newStr)-1]&0xC0 == 0x80 {
-		for i := len(newStr) - 1; i >= 0; i-- {
-			if newStr[i]&0xC0 == 0xC0 {
-				return newStr[:i] + TruncationSuffix
-			}
-		}
-	}
-
-	return newStr + TruncationSuffix
-}
-
 // FormatMessage formats a BSON document for logging. The document is truncated
 // to the given width.
-func FormatMessage(msg string, width uint) string {
+func FormatMessage(msg bsoncore.Document, width uint) string {
```

Comment:
We can clean up the usage of `FormatMessage` by updating the function signature to use `bson.Raw` instead of `bsoncore.Document`. Suggest the following changes:

First, extend the bson.Raw and bson.RawValue types with a `StringN` method:

```go
// bson/raw.go

func (r Raw) StringN(n int) string { return bsoncore.Document(r).StringN(n) }
```

```go
// bson/raw_value.go

func (rv RawValue) StringN(n int) string { return convertToCoreValue(rv).StringN(n) }
```

Suggest also renaming this function to `FormatDocument`:

```go
// internal/logger/logger.go

func FormatDocument(msg bson.Raw, width uint) string {
	if len(msg) == 0 {
		return "{}"
	}

	return msg.StringN(int(width)) + TruncationSuffix
}
```

Next, to avoid complex lines like this one:

```go
// x/mongo/driver/operation.go

formattedReply := logger.FormatMessage(bsoncore.BuildDocument(nil, bsoncore.AppendStringElement(nil, "key", info.cmdErr.Error())), op.Logger.MaxDocumentLength)
```

Suggest adding a `FormatString` function as well:

```go
// internal/logger/logger.go

func FormatString(str string, width uint) string {
	rawValue := bson.RawValue{
		Type:  bson.TypeString,
		Value: bsoncore.AppendString(nil, str),
	}

	return rawValue.StringN(int(width)) + TruncationSuffix
}
```

```go
// x/mongo/driver/operation.go

formattedReply := logger.FormatString(info.cmdErr.Error(), op.Logger.MaxDocumentLength)
```